bothan-lib/src/worker/websocket.rs

```diff
@@ -93,32 +99,48 @@ pub async fn start_polling<S, E1, E2, P, C>(
 }
 
 // If connect fails, the polling loop should be exited
-async fn connect<C, P, E1, E2>(connector: &C, ids: &[String], opts: &PollOptions) -> Option<P>
+async fn connect<C, P, E1, E2>(
+    connector: &C,
+    ids: &[String],
+    opts: &PollOptions,
+    metrics: WebSocketMetrics,
+) -> Option<P>
 where
     P: AssetInfoProvider<SubscriptionError = E1, PollingError = E2>,
     C: AssetInfoProviderConnector<Provider = P>,
 {
+    let start_time = chrono::Utc::now().timestamp_millis();
     let mut retry_count: u64 = 1;
 
-    let histogram = global::meter(opts.meter_name)
-        .u64_histogram("connection-attempts")
-        .with_unit("attempt")
-        .build();
-
     while retry_count <= opts.max_retry {
         if let Ok(mut provider) = connector.connect().await {
             if provider.subscribe(ids).await.is_ok() {
-                histogram.record(retry_count, &[KeyValue::new("success", true)]);
+                metrics.record_connection_duration(
+                    chrono::Utc::now().timestamp_millis() - start_time,
+                    opts.worker_name.clone(),
+                    ConnectionResult::Success,
+                );
+                metrics.increment_connections_total(
+                    opts.worker_name.clone(),
+                    ConnectionResult::Success,
+                );
+
                 return Some(provider);
             }
         }
 
-        histogram.record(retry_count, &[KeyValue::new("success", true)]);
         error!("failed to reconnect. current attempt: {}", retry_count);
 
         retry_count += 1;
         sleep(opts.reconnect_buffer).await;
     }
 
+    metrics.record_connection_duration(
+        chrono::Utc::now().timestamp_millis() - start_time,
+        opts.worker_name.clone(),
+        ConnectionResult::Failed,
+    );
+    metrics.increment_connections_total(opts.worker_name.clone(), ConnectionResult::Failed);
```

Comment:
im pretty sure we can use some sort of builder pattern or other method to keep the worker_name as part of the metric so we dont need to specify it here